primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+impl<T: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath> Index for T {}
+
+/// An internal node value type in a Merkle tree.
+pub trait NodeValue: Default + Eq + PartialEq + Copy + Clone + Debug {}
+impl<T: Default + Eq + PartialEq + Copy + Clone + Debug> NodeValue for T {}
+
+/// Merkle tree hash function
+pub trait DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    // Possible improvement: adding digest_element() and digest_index()
+    /// Digest a list of values
+    fn digest(data: &[T]) -> T;
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
+    /// Convert the given index to a vector of branch indices given tree height
+    /// and arity.
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize>;
+}
+
+impl<F: Into<BigUint> + Clone> ToTraversalPath for F {
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize> {
+        let mut pos: BigUint = <F as Into<BigUint>>::into(self.clone());
+        let mut ret = vec![];
+        for _i in 0..height {
+            ret.push((&pos % (arity as u64)).to_usize().unwrap());
+            pos /= arity as u64;
+        }
+        ret
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct MerkleCommitment<T: NodeValue> {
+    /// Root of a tree
+    pub root_value: T,
+    /// Height of a tree
+    pub height: usize,
+    /// Number of leaves in the tree
+    pub num_leaves: u64,
+}
+
+/// Basic functionalities for a merkle tree implementation. Abstracted as an
+/// accumulator for fixed-length array. Supports generate membership proof at a
+/// given position and verify a membership proof.
+pub trait MerkleTreeScheme: Sized {
+    /// Merkle tree element type
+    type Element: Element;
+    /// Hash algorithm used in merkle tree
+    type Digest: DigestAlgorithm<Self::Element, Self::Index, Self::NodeValue>;
+    /// Index type for this merkle tree
+    type Index: Index;
+    /// Internal and root node value
+    type NodeValue: NodeValue;
+    /// Merkle proof
+    type MembershipProof: Clone;
+    /// Batch proof
+    type BatchMembershipProof: Clone;
+
+    /// Tree arity
+    const ARITY: usize;
+
+    /// Construct a new merkle tree with given height from a data slice
+    fn from_elems(
+        height: usize,
+        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
+    ) -> Result<Self, PrimitivesError>;
+
+    /// Return the height of this merkle tree
+    fn height(&self) -> usize;
+    /// Return the maximum allowed number leaves
+    fn capacity(&self) -> BigUint;
+    /// Return the current number of leaves
+    fn num_leaves(&self) -> u64;
+
+    /// Return the current root value
+    fn root(&self) -> Self::NodeValue;
```

Comment:
Not sure this really belongs in the trait. For an abstract interface, `commitment` serves the purpose of giving the tree a concise identifier, and I'm not sure what other functionality the user would expect from `root` -- especially with the return value being an opaque type. If we were to get rid of this, and implement the change I suggested to make `Commitment` an associated type, we could remove the `NodeValue` associated type, since it would no longer appear in the interface.

Reply:
Good suggestion. Will do.